# Assesment Problems

In [22]:
# Import Libraries 
import numpy as np
import math

np.random.seed(50)

## Problem 1: Extending the Lady Tasting Tea

This experiment extends Fisher’s [*Lady Tasting Tea*](https://en.wikipedia.org/wiki/Lady_tasting_tea) design to compare the probability of perfectly identifying all cups in an 8-cup versus a 12-cup setup.
We will use both analytical calculation (`math.comb`) and a Monte Carlo simulation (`numpy`) to estimate these probabilities.

In [23]:
# Define Experiment Parameters.

# Original experiment.
# total cups.
no_cups_orig = 8
# milk-first cups.
no_cups_milk_first_orig = 4  
# tea-first cups.      
no_cups_tea_first_orig = 4           

# Extended experiment.
# total cups (extended).
no_cups_ext = 12
# milk-first cups (extended).                     
no_cups_milk_first_ext = 4  
# tea-first cups (extended).         
no_cups_tea_first_ext = 8           


In [24]:
# Number of ways of selecting 4 milk-first cups from 8 total cups.
ways_orig = math.comb(no_cups_orig, no_cups_milk_first_orig)
# Number of ways of selecting 4 milk-first cups from 12 total cups.
ways_ext = math.comb(no_cups_ext, no_cups_tea_first_ext)

# Show
ways_orig,ways_ext

(70, 495)

[Reference: Python math.comb() Method](https://www.geeksforgeeks.org/python/python-math-comb-method/)  

In [25]:
# Calculate exact probabilities.
prob_orig_exact = 1 / ways_orig
prob_ext_exact  = 1 / ways_ext

print(f"Original experiment exact probability:  {prob_orig_exact}")
print(f"Extended experiment exact probability:  {prob_ext_exact}")

Original experiment exact probability:  0.014285714285714285
Extended experiment exact probability:  0.00202020202020202


In [26]:
# ---- Monte Carlo simulation of random guesses ----

# Seed for reproducibility and consistent results
np.random.seed(50)

# Number of random guesses (10,000 gives stable results with minimal runtime)
# See: Practical Business Python – Monte Carlo Simulation with Python

n_shuffles = 10_000 

# --- Original experiment (8 cups) ---
# Counter to record how many times all cups are guessed correctly
hits_orig = 0

# Repeat the experiment n_shuffles times
for _ in range(n_shuffles):
    # Randomly select 4 cups (from 0–7) to represent the participant's guess
    guess = np.random.choice(no_cups_orig, size=no_cups_milk_first_orig, replace=False)
    
    # Assume the true milk-first cups are {0, 1, 2, 3}.
    # Compare the guessed cups with the true cups as sets.
    if set(guess) == set(range(no_cups_milk_first_orig)):
        # If the guess matches exactly, count it as one.
        hits_orig += 1

# Calculate the simulated probability of a perfect guess
prob_orig_sim = hits_orig / n_shuffles


# --- Extended experiment (12 cups) ---
# Counter to record how many times all cups are guessed correctly
hits_ext = 0

# Repeat the experiment n_shuffles times
for _ in range(n_shuffles):
    # Randomly select 4 cups (from 0–11) to represent the participant's guess
    guess = np.random.choice(no_cups_ext, size=no_cups_milk_first_ext, replace=False)
    
    # Assume the true milk-first cups are {0, 1, 2, 3}.
    # Compare the guessed cups with the true cups as sets.
    if set(guess) == set(range(no_cups_milk_first_ext)):
        # If the guess matches exactly, count it as one.
        hits_ext += 1

# Calculate the simulated probability of a perfect guess
prob_ext_sim = hits_ext / n_shuffles


# --- Display the results ---
print(f"Shuffles: {n_shuffles:,}")
print(f"Original  — exact probability: {prob_orig_exact:.6f}, simulated probability: {prob_orig_sim:.6f}")
print(f"Extended  — exact probability: {prob_ext_exact:.6f}, simulated probability: {prob_ext_sim:.6f}")


Shuffles: 10,000
Original  — exact probability: 0.014286, simulated probability: 0.015100
Extended  — exact probability: 0.002020, simulated probability: 0.002500


[Reference: Using Monte Carlo Simulation for Sampling in Python (Medium)](https://medium.com/%40kaanalperucan/using-monte-carlo-simulation-for-sampling-in-python-with-examples-7a9600dfb73e) 
[Reference: Monte Carlo Simulation – An In-Depth Tutorial (Towards AI)](https://towardsai.net/p/editorial/monte-carlo-simulation-an-in-depth-tutorial-with-python-bcf6eb7856c8)  
[Reference: Monte Carlo Simulation with Python (Practical Business Python)](https://pbpython.com/monte-carlo.html)  
[Reference: A Guide to Monte Carlo Simulation (Analytics Vidhya)](https://www.analyticsvidhya.com/blog/2021/07/a-guide-to-monte-carlo-simulation/)  
[Reference: NumPy random.choice() Documentation (DataCamp)](https://www.datacamp.com/doc/numpy/random-choice)  



### Interpretation
The simulated probabilities closely match the exact analytical results, confirming that the Monte Carlo approach accurately estimates the probability of a perfect guess.  
For the original 8-cup design, the probability of correctly identifying all cups by chance is approximately 0.0143 (1 in 70).  
For the extended 12-cup design, this drops to about 0.0020 (1 in 495), showing that the likelihood of a perfect random guess decreases  as the number of cups increases.  
Because the extended experiment makes a perfect guess so unlikely, one might reasonably consider relaxing the p-value threshold compared to the original setup, as even partial success would already be highly unlikely just by chance. 

[P-Value in Statistical Hypothesis Tests: What is it? (Statistics How to)](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/p-value/)  

---

## Problem 2: Normal distribution

Here we generate many samples from a standard normal distribution, N(0, 1), and compare two ways of calculating the standard deviation for each sample.

Population standard deviation (ddof = 0) divides by n and assumes the data represent the entire population. It is slightly biased when used on a sample because small samples tend to underestimate the true spread.

Sample standard deviation (ddof = 1) divides by n – 1 to correct that bias when estimating from a sample. This version gives a better estimate of the true population spread, especially for small sample sizes.

[Population vs. Sample Standard Deviation: When to Use Each (Statology)](https://www.statology.org/population-vs-sample-standard-deviation/)  

In [27]:
# Number of samples
N = 100000
# Size of each sample
n = 10

# Draw N samples of size n from a standard normal distribution, N(0, 1)
# Result is a 2D array with N rows and n columns
samples = np.random.normal(0, 1, (N, n))

# Calculate the standard deviation of each sample
# Population SD
sd_pop = np.std(samples, axis=1, ddof=0)
# Sample SD
sd_smp = np.std(samples, axis=1, ddof=1) 

[numpy.random.normal (NumPy)](https://numpy.org/doc/2.1/reference/random/generated/numpy.random.normal.html)  
[Normal Distribution in NumPy (GeeksforGeeks)](https://www.geeksforgeeks.org/numpy/normal-distribution-in-numpy/)  
[numpy.std (NumPy)](https://numpy.org/doc/2.1/reference/generated/numpy.std.html)  
[numpy.std() in Python (GeeksforGeeks)](https://www.geeksforgeeks.org/python/numpy-std-in-python/)  
[How to Compute Standard Deviation in NumPy (Spark By Examples)](https://sparkbyexamples.com/python/python-numpy-standard-deviation-function/)


In [28]:
np.mean(sd_pop), np.mean(sd_smp)

(0.9233056431863863, 0.9732496029852354)